File: 0ther-impl/PHP/xorfiles.php
#!/usr/bin/env php
<?php

const BUF = 524288;

/// Start processing CLI params ---->
if (count($argv) < 2 || count($argv) > 4) {
    echo "Usage:\n\$ xorfiles.php {first-file | stdin} second-file [output-file]\n\n";
    exit(22);
}

switch (count($argv)) {
    case 2:
        $firstFile = fopen('php://stdin', 'r');
        $secondFile = fopen($argv[1], 'r');
        $outputFile = fopen('php://stdout', 'w');
        break;
    case 3:
        if (file_exists($argv[2]) === true) {
            $firstFile = fopen($argv[1], 'r');
            $secondFile = fopen($argv[2], 'r');
            $outputFile = fopen('php://stdout', 'w');
        } else {
            $firstFile = fopen('php://stdin', 'r');
            $secondFile = fopen($argv[1], 'r');
            $outputFile = fopen($argv[2], 'w');
        }
        break;
    default:
            $firstFile = fopen($argv[1], 'r');
            $secondFile = fopen($argv[2], 'r');
            $outputFile = fopen($argv[3], 'w');
}

if ($firstFile === false || $secondFile === false || $outputFile === false) {
    echo "\nERROR! Can't open some specified files. Please check.\n\n";
    exit(2);
} /// <---- End processing CLI params

$firstUnit = '';
$secondUnit = '';
$result = '';

while(!feof($firstFile) && !feof($secondFile)) {

    $firstUnit = fread($firstFile, BUF);

    if (strlen($firstUnit) > 0) {
        $secondUnit = fread($secondFile, strlen($firstUnit));
        $result = $firstUnit ^ $secondUnit;
        if (fwrite($outputFile, $result, BUF) === false) {
            fwrite(STDERR, "\nERROR! Can't write data to the output-file (prehaps no space left). Exiting.\n");
            exit(5);
        }
    }
}

fclose($firstFile);
fclose($secondFile);
fclose($outputFile);

File: 0ther-impl/Python/xorfiles.py
#!/usr/bin/env python3

import sys
import os

BUF = 524288

# Start processing CLI params
if len(sys.argv) < 2 or len(sys.argv) > 4:
    print("Usage:\n$ xorfiles.py {first-file | stdin} second-file [output-file]\n")
    sys.exit(22)

if len(sys.argv) == 2:
    first_file = sys.stdin.buffer
    second_file = open(sys.argv[1], 'rb')
    output_file = sys.stdout.buffer
elif len(sys.argv) == 3:
    if os.path.exists(sys.argv[2]):
        first_file = open(sys.argv[1], 'rb')
        second_file = open(sys.argv[2], 'rb')
        output_file = sys.stdout.buffer
    else:
        first_file = sys.stdin.buffer
        second_file = open(sys.argv[1], 'rb')
        output_file = open(sys.argv[2], 'wb')
else:
    first_file = open(sys.argv[1], 'rb')
    second_file = open(sys.argv[2], 'rb')
    output_file = open(sys.argv[3], 'wb')

if first_file is None or second_file is None or output_file is None:
    print("\nERROR! Can't open some specified files. Please check.\n")
    sys.exit(2)

while True:

    chunk1 = first_file.read(BUF)
    chunk2 = second_file.read(BUF)

    if not chunk1 or not chunk2:
        break

    length = min(len(chunk1), len(chunk2))

    result = bytes(a ^ b for a, b in zip(chunk1[:length], chunk2[:length]))

    if output_file.write(result) is None:
        print("\nERROR! Can't write data to the output-file (perhaps no space left). Exiting.", file=sys.stderr)
        sys.exit(5)

first_file.close()
second_file.close()
output_file.close()
